Add Secure Private Access Support (#44)
aklivity/zilla-plus-aws-templates: amazon-msk/cdk/secure-private-access/README.md

### List the inbound rules for MSK Serverless Security Group

```bash
aws ec2 describe-security-groups \
  --group-ids $(aws kafka list-clusters-v2 \
    --cluster-type SERVERLESS \
    --query "ClusterInfoList[?ClusterArn=='<msk-serverless-arn>'].Serverless.VpcConfigs[].SecurityGroupIds[]" \
    --output text) \
  --query "SecurityGroups[].IpPermissions" \
  --output json
```

## Required CDK Context Variables

You can set these variables in your `context` in `cdk.json` file under `zilla-plus` object.

### `vpcId`: VPC ID
The VPC ID where the MSK Serverless lives. 

```bash
aws ec2 describe-subnets --subnet-ids $(aws kafka describe-cluster --cluster-arn <msk-cluster-arn> --query "ClusterInfo.BrokerNodeGroupInfo.ClientSubnets[0]" --output text) --query "Subnets[0].VpcId" --output text
```

### `msk` related variables

```json
    "msk":
    {
        "servers": "<Bootstrap Servers of your MSK Serverless>",
        "subnetIds": "<Subnet IDs of your MSK Serverless>"
    }
```

#### `servers`: MSK Bootstrap Servers and Authentication Method

To get the bootstrap servers of the MSK cluster run:

```bash
aws kafka get-bootstrap-brokers \
    --cluster-arn <msk-serverless-arn> \

>     --query '{BootstrapBrokerStringTls: BootstrapBrokerStringTls, BootstrapBrokerStringSaslScram: BootstrapBrokerStringSaslScram, BootstrapBrokerStringSaslIam: BootstrapBrokerStringSaslIam}' \

Reviewer: IAM only, right?

```

Once your stack is deployed, note down the VPC Endpoint Service Id and the VPC Endpoint Service Name, as you'll need this in the following steps when you add the VPC Endpoint from the client VPC.

## Connect to your MSK Serverless from different VPC
### Add a VPC endpoint in the client VPC
You need to add a VPC Endpoint pointing to the Zilla Plus's VPC Endpoint Service in your client VPC. 

```bash
aws ec2 create-vpc-endpoint \
  --vpc-id <Client VPC ID> \
  --service-name <VPC Endpoint Service Name> \
  --subnet-ids <Client Subnets> \
  --vpc-endpoint-type Interface
```

Make sure to save the `VpcEndpointId` and the the first DNS Entry `DNSName`.

### Accept the VPC Endpoint in your VPC Endpoint Service
```bash
aws ec2 accept-vpc-endpoint-connections \
  --service-id <VPC Endpoint Service ID> \
  --vpc-endpoint-ids <VPC Endpoint ID>
```

### Add Route 53 Private Hosted Zone
For your client machine to be able to resolve the custom wilcard DNS configured for Zilla, you need to add a Route 53 Private Hosted Zone, with a `CNAME` record that resolved the custom domain to your previously created VPC Endpoint DNS name.

```bash
aws route53 create-hosted-zone \

>   --name example.aklivity.io \

Reviewer: Use example with `us-east-1` prefix, as including the region is best practice so we want to lead them there.

  --vpc VPCRegion=us-east-1,VPCId=vpc-09cb6e2141bdb7b37 \
  --caller-reference <unique caller id> \
  --hosted-zone-config PrivateZone=true 
```

Use the HostedZone.Id in your following command:

```bash
aws route53 change-resource-record-sets \
  --hosted-zone-id /hostedzone/Z0357410VJ4A0M2XZFAK \
  --change-batch '{
    "Changes": [
      {
        "Action": "CREATE",
        "ResourceRecordSet": {
          "Name": "*.us-east-1.example.aklivity.io",
          "Type": "CNAME",
          "TTL": 60,
          "ResourceRecords": [
            {
              "Value": "vpce-0ee315e6fc87716da-ia1ca1zt.vpce-svc-0816be0e665a42f9a.us-east-1.vpce.amazonaws.com"
            }

>           ]

Reviewer: Use ALIAS record type instead.

You can either choose to use `AWS_MSK_IAM` or `OAUTHBEARER`.

##### client.properties for `AWS_MSK_IAM`

```text
security.protocol=SASL_SSL
sasl.mechanism=AWS_MSK_IAM
sasl.jaas.config=software.amazon.msk.auth.iam.IAMLoginModule required;
sasl.client.callback.handler.class=software.amazon.msk.auth.iam.IAMClientCallbackHandler
```

##### client.properties for `OAUTHBEARER`

```text
security.protocol=SASL_SSL
sasl.mechanism=OAUTHBEARER
sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required;
sasl.login.callback.handler.class=software.amazon.msk.auth.iam.IAMOAuthBearerLoginCallbackHandler
sasl.client.callback.handler.class=software.amazon.msk.auth.iam.IAMOAuthBearerLoginCallbackHandler
```

TODO:
- add iam role readme
- add steps to add more supported regions, point to the filed issue that limits us to automate this

### Test the Kafka Client

This verifies internet connectivity to your MSK Serverless via Zilla Plus.

We can now verify that the Kafka client can successfully communicate with your MSK Serverless from your EC2 instance running in a different VPC to create a topic, then publish and subscribe to the same topic.

> If using the wildcard DNS pattern `*.example.aklivity.io`, then we use the following server name for the Kafka client:

Reviewer: ```suggestion
If using the wildcard DNS pattern `*.us-east-1.example.aklivity.io`, then we use the following server name for the Kafka client:
```

```text

> boot.example.aklivity.io:9098

Reviewer: ```suggestion
boot.us-east-1.example.aklivity.io:9098
```

```

Replace these bootstrap server names accordingly for your own custom wildcard DNS pattern.

#### Create a Topic

Use the Kafka client to create a topic called zilla-proxy-test, replacing <tls-bootstrap-server-names> in the command below with the TLS proxy names of your Zilla proxy:

```bash

> bin/kafka-topics.sh --create --topic zilla-proxy-test --partitions 3 --replication-factor 2 --command-config client.properties --bootstrap-server <tls-bootstrap-server-names>

Reviewer: ```suggestion
bin/kafka-topics.sh --create \
    --topic zilla-proxy-test \
    --partitions 3 \
    --replication-factor 2 \
    --command-config client.properties \
    --bootstrap-server boot.us-east-1.example.aklivity.io:9098
```

Use the example, and change the instruction to replace `boot.us-east-1.example.aklivity.io:9098` instead.

```bash
aws route53 create-hosted-zone \

>   --name us-east-1.example.aklivity.io \

Reviewer: ```suggestion
  --name example.aklivity.io \
```

sasl.client.callback.handler.class=software.amazon.msk.auth.iam.IAMOAuthBearerLoginCallbackHandler
```

### Test the Kafka Client

This verifies connectivity to your MSK Serverless via Zilla Plus over a different VPC.

We can now verify that the Kafka client can successfully communicate with your MSK Serverless from your EC2 instance running in a different VPC to create a topic, then produce and subscribe to the same topic.

> If using the wildcard DNS pattern `*.us-east-1.example.aklivity.io`, then we use the following server name for the Kafka client:

Reviewer: ```suggestion
If using the wildcard DNS pattern `*.example.aklivity.io`, then we use the following server name for the Kafka client:
```

```text

> boot.us-east-1.example.aklivity.io:9098

Reviewer: ```suggestion
boot.example.aklivity.io:9098
```

```

Replace these bootstrap server names accordingly for your own custom wildcard DNS pattern.

#### Create a Topic

> Use the Kafka client to create a topic called zilla-proxy-test, replacing <bootstrap-server-names> in the command below with the proxy names of your Zilla proxy:

Reviewer: ```suggestion
Use the Kafka client to create a topic called zilla-proxy-test, updating `boot.example.aklivity.io:9098` in the command below to use the custom domain of your Zilla proxy:
```

```bash
bin/kafka-topics.sh --create \
    --topic zilla-proxy-test \
    --partitions 3 \
    --replication-factor 2 \
    --command-config client.properties \

>     --bootstrap-server boot.us-east-1.example.aklivity.io:9098

Reviewer: ```suggestion
    --bootstrap-server boot.example.aklivity.io:9098
```
